src/azul/openapi.py

```diff
@@ -0,0 +1,113 @@
+    """
+    Add description to decorated function
+    >>> @spec({'a': 'b'})
+    ... def foo():
+    ...     pass
+    ...
+
+    >>> foo.api_spec == {'a': 'b'}
+    True
+    """
+
+    def spec_adder(func):
+        func.api_spec = description
+        return func
+    return spec_adder
+
+
+def annotated_specs(gateway_id, app, api_info) -> JSON:
+    """
+    Finds all routes in app that are decorated with @spec and adds this information
+    into the api spec downloaded from API Gateway.
+    :param gateway_id: API Gateway ID corresponding to the Chalice app
+    :param app: App with annotated routes
+    :param api_info: Top level OpenAPI info, definitions, etc.
+    :return: The annotated specifications
+    """
+    specs = aws.api_gateway_export(gateway_id)
+    clean_specs(specs)
+    # TODO: if we want to set version, override needs to be true.
+    specs = merge_dicts(specs, api_info)
+    docs = get_doc_specs(app)
+    for path, doc_spec in docs.items():
+        for verb, description in specs['paths'][path].items():
+            description = merge_dicts(description, doc_spec)
+            specs['paths'][path][verb] = description
+    return specs
+
+
+def clean_specs(specs):
+    """
+    Adjust specs from API Gateway so they pass linting
+    """
+    # Filter out 'options' since it causes linting errors
+    for path in specs['paths']:
+        specs['paths'][path].pop('options', None)
+
+
+def get_doc_specs(app):
+    docs = {}
+    for route, entries in app.routes.items():
+        func = next(iter(entries.values())).view_function
+        if hasattr(func, 'api_spec'):
+            docs[route] = func.api_spec
+    return docs
+
+
+def merge_dicts(d1: MutableMapping, d2: MutableMapping, mutate: bool = False, override: bool = False):
+    """
+    Merge two dictionaries deeply such that:
+    - collisions raise an error,
+    - lists are appended, and
+    - child dictionaries are recursively merged
+    If override is true, d2 overrides d1 on collisions.
+    >>> merge_dicts({1: 'a'}, {2: 'a'})
+    {1: 'a', 2: 'a'}
+    >>> merge_dicts({1: 1}, {1: 2})
+    Traceback (most recent call last):
+    ...
+    ValueError: Cannot merge path '1' with values: 1, 2
+    >>> merge_dicts({1: 1}, {1: 2}, override=True)
+    {1: 2}
+    >>> merge_dicts({1: [1, 2]}, {1: [2, 3]})
+    {1: [1, 2, 2, 3]}
+    >>> merge_dicts({1: {'a': 'z', 'c': [1]}}, {1: {'b': 'y', 'c': [2]}})
+    {1: {'a': 'z', 'c': [1, 2], 'b': 'y'}}
+    >>> merge_dicts({1: {'a': 'z'}}, {1: {'a': 'y'}})
+    Traceback (most recent call last):
+    ...
+    ValueError: Cannot merge path '1.a' with values: z, y
+    >>> merge_dicts({1: {'a': 'z'}}, {1: {'a': 'y'}}, override=True)
+    {1: {'a': 'y'}}
+    """
+    if mutate:
+        d1 = copy.deepcopy(d1)
+        d2 = copy.deepcopy(d2)
+    return recursive_merge_dicts(d1, d2, [], override)
+
+
+def recursive_merge_dicts(merged: MutableMapping, d2: MutableMapping, path: List[str], override: bool):
```

Comment:
If a method is solely for internal use, prefix its name with underscore. It helps communicating what's intended to be used from the outside and what should be considered internal.

Reply:
Good point.